R/bm-array-altrep-materialization.R

```diff
@@ -0,0 +1,123 @@
+#' Benchmark for materializing an altrep Arrow array
+#'
+#' This flexes a lower level conversion to R data structures from Arrow data structures.
+#'
+#' @section Parameters:
+#' * `source` A known-file id to use (it will be read in to a data.frame first)
+#' * `exclude_nulls` Logical. Remove any columns with any `NULL`s or `NA`s in them?
+#' * `altrep` Logical. Use altrep storage for vectors?
+#' * `subset_indices` Length-one list of vector to use to subset rows of source.
+#'
+#' @importFrom purrr transpose
+#' @export
+array_altrep_materialization <- Benchmark(
+  "array_altrep_materialization",
+
+  setup = function(source = names(known_sources),
+                   exclude_nulls = FALSE,
+                   altrep = TRUE,
+                   subset_indices = list(1:10)) {
+    stopifnot(
+      is.logical(exclude_nulls),
+      is.logical(altrep)
+    )
+    source <- match.arg(source, names(known_sources))
+    subset_indices <- subset_indices[[1]]
+
+    options(arrow.use_altrep = altrep)
+    path <- ensure_format(source, format = "parquet", compression = "snappy")
+
+    # exclude non-altrep types
+    pq <- arrow::ParquetFileReader$create(path)
+    table_arrow_types <- lapply(pq$GetSchema()$fields, function(x) x$type)
+    arrow_altrep_types <- list(
+      arrow::float64(),
+      arrow::int32(),
+      arrow::utf8(),
+      arrow::large_utf8(),
+      arrow::dictionary()
+    )
+    table_arrow_altrep_type_mat <- outer(
+      table_arrow_types,
+      arrow_altrep_types,
+      Vectorize(function(x, y) x == y)
+    )
+    is_arrow_altrep_type <- apply(table_arrow_altrep_type_mat, 1, any)
+    table <- pq$ReadTable(which(is_arrow_altrep_type) - 1L)
+
+    if (exclude_nulls) {
+      cols_without_nulls <- vapply(
+        colnames(table),
+        function(x) table[[x]]$null_count == 0,
+        logical(1L)
+      )
+      table <- table[which(cols_without_nulls)]
+    }
+
+    # extract the arrays
+    arrays <- as.list(table)
+    array_lengths <- lengths(arrays)
+
+
+    # altrep checking
```

Comment:
```suggestion

    # altrep checking
```

Almost asinine whitespace nit